File: polyglot/java/cve_ingestion.java
package polyglot.java;

import java.io.*;
import java.net.http.HttpClient;
import java.net.URI;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * CVE Ingestion Module for vulntimeline.
 * Parses, validates, deduplicates, and normalizes CVE records from multiple sources.
 */
public class CveIngestion {

    // NVD default feed URL
    private static final String DEFAULT_NVD_URL = "https://services.nvd.nist.gov/rest/json/cves/2.0";
    
    // Deduplication key: normalized CVE ID (lowercase, trimmed)
    private static final Pattern CVE_PATTERN = Pattern.compile(
        "^CVE-\\d{4}-[A-Z]{1,6}[0-9]{1,3}$", 
        Pattern.CASE_INSENSITIVE | Pattern.MULTILINE);

    /**
     * Normalized CVE record with all extracted metadata.
     */
    public static class CveRecord {
        private final String cveId;
        private final Instant publishedDate;
        private final Instant lastModified;
        private final int severityScore; // NVD CVSS v3.x or v4.0
        private final String cvssVersion;
        private final List<String> affectedProducts;
        private final Map<String, String> platforms = new LinkedHashMap<>();
        private final String description;
        private final boolean isRetired;

        public CveRecord(String cveId, Instant publishedDate, int severityScore, 
                        String cvssVersion, List<String> affectedProducts,
                        Map<String, String> platforms, String description, boolean isRetired) {
            this.cveId = normalizeCveId(cveId);
            this.publishedDate = publishedDate;
            this.lastModified = Instant.now();
            this.severityScore = severityScore;
            this.cvssVersion = cvssVersion != null ? cvssVersion : "3.1";
            this.affectedProducts = affectedProducts;
            this.platforms = platforms;
            this.description = description;
            this.isRetired = isRetired;
        }

        public String getNormalizedId() { return cveId; }
        public Instant getPublishedDate() { return publishedDate; }
        public int getSeverityScore() { return severityScore; }
        public List<String> getAffectedProducts() { return affectedProducts; }
        public Map<String, String> getPlatforms() { return platforms; }
        public boolean isRetired() { return isRetired; }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof CveRecord)) return false;
            CveRecord that = (CveRecord) o;
            return Objects.equals(cveId, that.cveId);
        }

        @Override
        public int hashCode() {
            return cveId != null ? cveId.hashCode() : 0;
        }

        @Override
        public String toString() {
            return "CveRecord{" +
                   "cveId='" + cveId + '\'' +
                   ", severity=" + severityScore +
                   ", published=" + publishedDate +
                   '}';
        }
    }

    /**
     * Ingestion configuration.
     */
    public static class Config {
        private String nvdUrl = DEFAULT_NVD_URL;
        private int batchSize = 100;
        private boolean autoRetry = true;
        private int maxRetries = 3;
        private long timeoutMs = 30_000;

        public static Config defaultConfig() {
            return new Config();
        }

        public String getNvdUrl() { return nvdUrl; }
        public void setNvdUrl(String url) { this.nvdUrl = url; }
        public int getBatchSize() { return batchSize; }
        public boolean isAutoRetry() { return autoRetry; }
        public long getTimeoutMs() { return timeoutMs; }

        @Override
        public String toString() {
            return "Config{" +
                   "nvdUrl='" + nvdUrl + '\'' +
                   ", batchSize=" + batchSize +
                   '}';
        }
    }

    /**
     * Thread-safe deduplication store.
     */
    private static class DedupStore {
        private final Map<String, CveRecord> records = new ConcurrentHashMap<>();
        private final Set<String> pending = ConcurrentHashMap.newKeySet();

        public void add(CveRecord record) {
            if (record == null || record.getNormalizedId() == null) return;
            
            String key = record.getNormalizedId().toUpperCase();
            CveRecord existing = records.get(key);
            
            // Keep the one with more recent published date
            if (existing != null && 
                Objects.equals(existing.getPublishedDate(), record.getPublishedDate())) {
                records.put(key, record);
            } else if (existing == null || 
                       record.getPublishedDate().isAfter(existing.getPublishedDate())) {
                records.put(key, record);
            }
        }

        public List<CveRecord> getAll() {
            return new ArrayList<>(records.values());
        }

        public int size() {
            return records.size();
        }

        public void clear() {
            records.clear();
            pending.clear();
        }
    }

    /**
     * Main ingestion engine.
     */
    private static class IngestionEngine {
        private final Config config;
        private final DedupStore store;
        private final HttpClient client = HttpClient.newHttpClient();

        public IngestionEngine(Config config, DedupStore store) {
            this.config = Objects.requireNonNull(config);
            this.store = Objects.requireNonNull(store);
        }

        /**
         * Fetches CVE records from NVD API.
         */
        private List<CveRecord> fetchFromNvd(String cveId, int maxResults) throws IOException {
            String url = config.getNvdUrl() + "?cveId=" + UriEncode(cveId);
            
            try (var response = client.send(
                HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .GET()
                    .header("Accept", "application/json")
                    .timeout(java.time.Duration.ofMillis(config.getTimeoutMs()))
                    .build(),
                BodyHandlers.ofString()
            )) {

                if (response.statusCode() == 200) {
                    return parseNvdResponse(response.body());
                } else if (response.statusCode() >= 400 && response.statusCode() < 500) {
                    throw new IOException("HTTP " + response.statusCode() + ": " + response.body());
                } else {
                    // Retry for transient errors
                    return retryFetch(cveId, maxResults);
                }
            }
        }

        private List<CveRecord> parseNvdResponse(String json) throws IOException {
            try (var reader = new StringReader(json)) {
                JsonParser parser = new JsonParser();
                
                // NVD 2.0 response structure:
                // { "items": [ { "cve": { "id": "...", "published": "...", ... } }, ... ] }
                
                var items = parser.parse(reader, CveItem.class);
                List<CveRecord> result = new ArrayList<>();

                for (var item : items) {
                    if (!item.getCve().isPresent()) continue;

                    var cveData = item.getCve().get();
                    
                    // Extract timestamps
                    Instant published = parseTimestamp(cveData.getPublished());
                    Instant modified = parseTimestamp(cveData.getLastModified());

                    // Parse CVSS score from NVD fields
                    int severityScore = 0;
                    String cvssVersion = "3.1";
                    
                    if (cveData.getCvssV3_1() != null) {
                        severityScore = parseCvssScore(cveData.getCvssV3_1());
                        cvssVersion = "3.1";
                    } else if (cveData.getCvssV4_0() != null) {
                        severityScore = parseCvssScore(cveData.getCvssV4_0());
                        cvssVersion = "4.0";
                    }

                    // Extract affected products from NVD relationships
                    List<String> affectedProducts = new ArrayList<>();
                    if (cveData.getAffectedProducts() != null) {
                        for (var prod : cveData.getAffectedProducts()) {
                            String name = prod.getName();
                            String platform = prod.getPlatforms() != null ? 
                                prod.getPlatforms().get(0) : "Unknown";
                            
                            if (!name.isEmpty() && !platform.equals("Unknown")) {
                                affectedProducts.add(name + " (" + platform + ")");
                            }
                        }
                    }

                    // Build description summary
                    StringBuilder desc = new StringBuilder();
                    if (cveData.getDescription() != null) {
                        desc.append(cveData.getDescription().getSummary());
                    }
                    if (cveData.getReferences() != null && !cveData.getReferences().isEmpty()) {
                        desc.append(" | ").append(cveData.getReferences().size()).append(" refs");
                    }

                    // Check if retired
                    boolean isRetired = cveData.isRetired();

                    result.add(new CveRecord(
                        cveData.getId(),
                        published,
                        severityScore,
                        cvssVersion,
                        affectedProducts.isEmpty() ? null : affectedProducts,
                        new LinkedHashMap<>(), // platforms map (can be expanded)
                        desc.toString().trim(),
                        isRetired
                    ));
                }

                return result;
            }
        }

        private String parseTimestamp(String timestampStr) {
            if (timestampStr == null || timestampStr.isEmpty()) {
                return Instant.now();
            }

            // NVD uses ISO 8601 format: "2024-01-15T10:30:00.000Z"
            try {
                DateTimeFormatter formatter = DateTimeFormatter.ISO_INSTANT;
                return Instant.from(formatter.parse(timestampStr));
            } catch (Exception e) {
                // Fallback: assume epoch seconds if ISO parsing fails
                try {
                    long epochSeconds = Long.parseLong(timestampStr);
                    return Instant.ofEpochSecond(epochSeconds);
                } catch (NumberFormatException ex) {
                    return Instant.now();
                }
            }
        }

        private int parseCvssScore(CvssData cvssData) {
            if (cvssData == null || cvssData.getVector() == null) {
                return 0;
            }

            // CVSS v3.x uses a numeric score in the vector string
            String vector = cvssData.getVector();
            
            // Extract numeric score from "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:C/C:H/I:H/A:H"
            Pattern scorePattern = Pattern.compile("(\\d+\\.?\\d*)");
            Matcher matcher = scorePattern.matcher(vector);
            
            if (matcher.find()) {
                try {
                    return Integer.parseInt(matcher.group(1));
                } catch (NumberFormatException e) {
                    // Return 0 on parse failure
                }
            }

            // Fallback: use severity string mapping
            String severity = cvssData.getSeverity();
            if ("HIGH".equalsIgnoreCase(severity)) return 7;
            else if ("MEDIUM".equalsIgnoreCase(severity)) return 4;
            else if ("LOW".equalsIgnoreCase(severity)) return 1;
            
            return 0;
        }

        private String normalizeCveId(String id) {
            if (id == null || id.isEmpty()) return "UNKNOWN";
            
            // Remove leading/trailing whitespace, convert to uppercase
            id = id.trim().toUpperCase();
            
            // Handle common variations: CVE-2024-1234 vs CVE_2024_1234
            id = id.replace('_', '-');
            
            return id;
        }

        private String UriEncode(String str) {
            try {
                return java.net.URLEncoder.encode(str, "UTF-8");
            } catch (UnsupportedEncodingException e) {
                // UTF-8 should always be supported
                return str;
            }
        }

        private List<CveRecord> retryFetch(String cveId, int maxResults) throws IOException {
            if (!config.isAutoRetry()) {
                throw new IOException("Auto-retry disabled");
            }

            for (int attempt = 0; attempt < config.getMaxRetries(); attempt++) {
                try {
                    return fetchFromNvd(cveId, maxResults);
                } catch (IOException e) {
                    if (attempt == config.getMaxRetries() - 1) {
                        throw new IOException("Failed after " + (attempt + 1) + 
                                           " attempts: " + e.getMessage(), e);
                    }
                    
                    // Exponential backoff
                    long delay = Math.pow(2, attempt) * 500;
                    Thread.sleep(delay);
                }
            }
            
            return new ArrayList<>();
    }

        /**
         * Main ingestion entry point.
         */
        public List<CveRecord> ingest(String cveIdOrFile) throws IOException {
            if (cveIdOrFile == null || cveIdOrFile.isEmpty()) {
                throw new IllegalArgumentException("CVE ID or file path required");
            }

            // Check if it's a file path
            Path inputPath = Path.of(cveIdOrFile);
            boolean isFile = Files.exists(inputPath) && Files.isRegularFile(inputPath);

            List<CveRecord> results;

            if (isFile) {
                // Parse from JSON/XML file
                try (var reader = Files.newBufferedReader(inputPath)) {
                    JsonParser parser = new JsonParser();
                    var items = parser.parse(reader, CveItem.class);
                    
                    results = new ArrayList<>();
                    for (var item : items) {
                        if (!item.getCve().isPresent()) continue;
                        
                        var cveData = item.getCve().get();
                        Instant published = parseTimestamp(cveData.getPublished());
                        
                        // Simple file-based parsing for flexibility
                        int severityScore = 0;
                        String cvssVersion = "3.1";
                        
                        if (cveData.getCvssV3_1() != null) {
                            severityScore = parseCvssScore(cveData.getCvssV3_1());
                            cvssVersion = "3.1";
                        }

                        results.add(new CveRecord(
                            cveData.getId(),
                            published,
                            severityScore,
                            cvssVersion,
                            null, // affected products from file
                            new LinkedHashMap<>(),
                            cveData.getDescription() != null ? 
                                cveData.getDescription().getSummary() : "",
                            false
                        ));
                    }
                }
            } else {
                // Fetch from NVD API
                results = fetchFromNvd(cveIdOrFile, 10);
            }

            return results;
        }

        public void ingestBatch(List<String> cveIds) throws IOException {
            if (cveIds == null || cveIds.isEmpty()) {
                throw new IllegalArgumentException("CVE ID list required");
            }

            List<CveRecord> allRecords = new ArrayList<>();
            
            for (String cveId : cveIds) {
                // Batch fetch with parallel processing
                try {
                    var records = fetchFromNvd(cveId, 10);
                    allRecords.addAll(records);
                    
                    if (!config.isAutoRetry() && !records.isEmpty()) {
                        break; // Stop after first successful batch
                    }
                } catch (IOException e) {
                    System.err.println("Warning: Failed to fetch " + cveId + ": " + e.getMessage());
                }

                // Small delay between requests to avoid rate limiting
                Thread.sleep(100);
            }

            return allRecords;
        }

        public void ingestFromStream(InputStream inputStream) throws IOException {
            try (var reader = new InputStreamReader(inputStream)) {
                JsonParser parser = new JsonParser();
                var items = parser.parse(reader, CveItem.class);
                
                List<CveRecord> records = new ArrayList<>();
                for (var item : items) {
                    if (!item.getCve().isPresent()) continue;
                    
                    var cveData = item.getCve().get();
                    Instant published = parseTimestamp(cveData.getPublished());
                    int severityScore = 0;
                    
                    if (cveData.getCvssV3_1() != null) {
                        severityScore = parseCvssScore(cveData.getCvssV3_1());
                    }

                    records.add(new CveRecord(
                        cveData.getId(),
                        published,
                        severityScore,
                        "3.1",
                        null,
                        new LinkedHashMap<>(),
                        cveData.getDescription() != null ? 
                            cveData.getDescription().getSummary() : "",
                        false
                    ));
                }

                return records;
            }
        }
    }

    /**
     * JSON parser supporting NVD 2.0 response format.
     */
    private static class JsonParser {
        
        public <T> T parse(Reader reader, Class<T> clazz) throws IOException {
            // Simple recursive descent parser for NVD structure
            try (var scanner = new Scanner(reader)) {

File: polyglot/go/cve_ingestion.go
//go:build !noentrypoint
// +build !noentrypoint

package cve

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"
)

// Config holds ingestion settings.
type Config struct {
	BatchSize    int           `json:"batch_size"`
	RetryDelay   time.Duration `json:"retry_delay"`
	Timeout       time.Duration `json:"timeout"`
	MaxRetries    int           `json:"max_retries"`
	ParallelJobs  int           `json:"parallel_jobs"`
	StrictMode    bool          `json:"strict_mode"`
	LogLevel      string        `json:"log_level"`
	TempDir       string        `json:"temp_dir"`
}

// DefaultConfig returns sensible defaults.
func DefaultConfig() Config {
	return Config{
		BatchSize:    100,
		RetryDelay:   time.Second * 5,
		Timeout:      time.Minute,
		MaxRetries:   3,
		ParallelJobs: 4,
		StrictMode:   true,
		LogLevel:     "info",
		TempDir:      "/tmp/vulntimeline",
	}
}

// CVE represents a normalized vulnerability record.
type CVE struct {
	ID              string          `json:"id"`
	PublishedDate   time.Time       `json:"published_date"`
	UpdatedDate     time.Time       `json:"updated_date"`
	CVSSv3Score     float64         `json:"cvss_v3_score"`
	CVSSv2Score     float64         `json:"cvss_v2_score"`
	Summary         string          `json:"summary"`
	Description     string          `json:"description"`
	AffectedSystems []AffectedEntry `json:"affected_systems"`
	References      []Reference     `json:"references"`
	Publishers      []Publisher     `json:"publishers"`
	CVSSv3Vector    string          `json:"cvss_v3_vector"`
	CVSSv2Vector    string          `json:"cvss_v2_vector"`
}

// AffectedEntry describes a specific affected system.
type AffectedEntry struct {
	Product       string   `json:"product"`
	VersionRange  string   `json:"version_range"`
	Severity      string   `json:"severity"`
	CVSSv3Score   float64  `json:"cvss_v3_score,omitempty"`
	CVSSv2Score   float64  `json:"cvss_v2_score,omitempty"`
	PublishedDate time.Time `json:"published_date"`
}

// Reference holds a URL or identifier reference.
type Reference struct {
	URL       string    `json:"url"`
	Type      string    `json:"type"` // e.g., "advisory", "patch"
	Source    string    `json:"source"`
	Published time.Time `json:"published_date,omitempty"`
}

// Publisher identifies the organization that published the CVE.
type Publisher struct {
	Name       string    `json:"name"`
	URL        string    `json:"url"`
	Category   string    `json:"category"` // e.g., "CISA", "NVD"
	ContactURL string    `json:"contact_url,omitempty"`
}

// SourceResult tracks ingestion results per source.
type SourceResult struct {
	Name         string
	Count        int
	Errors       []error
	Duplicates   int
	Skipped      int
	LastModified time.Time
}

// IngestionResult aggregates all results.
type IngestionResult struct {
	TotalProcessed int
	TotalValid     int
	TotalInvalid   int
	TotalDuplicates int
	Results        map[string]*SourceResult
	Errors         []error
	StartTime      time.Time
	EndTime        time.Time
}

// NVDJSON represents the NVD JSON structure.
type NVDJSON struct {
	Metadata struct {
		Timestamp    string `json:"timestamp"`
		Version      string `json:"version"`
		UpdatedDate  string `json:"updated_date"`
		PublishedDate string `json:"published_date"`
	} `json:"metadata"`
	CVEItems []struct {
		ID              string     `json:"cve_id"`
		State           string     `json:"state"`
		PublishedDate   string     `json:"published_date"`
		UpdatedDate     string     `json:"updated_date"`
		VendorIDs       []string   `json:"vendor_ids"`
		Desc            struct {
			Value    string `json:"value"`
			Language string `json:"lang,omitempty"`
		} `json:"desc"`
		Summary         string     `json:"summary"`
		CvssData        []struct {
			Version   string  `json:"version"`
			VectorStr string  `json:"vectorString"`
			Score     float64 `json:"score,omitempty"`
		} `json:"cvss_data"`
		AffectedProducts []struct {
			Vendor    string `json:"vendor"`
			Product   string `json:"product"`
			Version   string `json:"version"`
			CvssData  []struct {
				Version   string  `json:"version"`
				VectorStr string  `json:"vectorString"`
				Score     float64 `json:"score,omitempty"`
			} `json:"cvss_data"`
		} `json:"affected_products"`
		References []struct {
			URL       string `json:"url"`
			Type      string `json:"ref_source"`
			Description struct {
				Value    string `json:"value"`
				Language string `json:"lang,omitempty"`
			} `json:"desc"`
		} `json:"references"`
		Publishers []struct {
			Name       string `json:"name"`
			URL        string `json:"url"`
			Category   string `json:"category"`
			ContactURL string `json:"contact_url,omitempty"`
		} `json:"publishers"`
	} `json:"cve_items"`
}

// CISAXML represents a simplified CISA XML structure.
type CISAXML struct {
	XMLName    string `xml:"CISACVE"`
	Metadata   struct {
		Timestamp    string `xml:"timestamp"`
		Version      string `xml:"version"`
		UpdatedDate  string `xml:"updated_date"`
		PublishedDate string `xml:"published_date"`
	} `xml:"metadata"`
	CVEItems []struct {
		ID              string     `xml:"cve_id"`
		State           string     `xml:"state"`
		PublishedDate   string     `xml:"published_date"`
		UpdatedDate     string     `xml:"updated_date"`
		VendorIDs       []string   `xml:"vendor_ids"`
		Desc            struct {
			Value    string `xml:",chardata"`
			Language string `xml:"lang,omitempty"`
		} `xml:"desc"`
		Summary         string     `xml:"summary"`
		CvssData        []struct {
			Version   string  `xml:"version"`
			VectorStr string  `xml:"vectorString"`
			Score      float64 `xml:"score,omitempty"`
		} `xml:"cvss_data"`
		AffectedProducts []struct {
			Vendor    string `xml:"vendor"`
			Product   string `xml:"product"`
			Version   string `xml:"version"`
			CvssData  []struct {
				Version   string  `xml:"version"`
				VectorStr string  `xml:"vectorString"`
				Score      float64 `xml:"score,omitempty"`
			} `xml:"cvss_data"`
		} `xml:"affected_products"`
		References []struct {
			URL       string `xml:"url"`
			Type      string `xml:"ref_source"`
			Description struct {
				Value    string `xml:",chardata"`
				Language string `xml:"lang,omitempty"`
			} `xml:"desc"`
		} `xml:"references"`
		Publishers []struct {
			Name       string `xml:"name"`
			URL        string `xml:"url"`
			Category   string `xml:"category"`
			ContactURL string `xml:"contact_url,omitempty"`
		} `xml:"publishers"`
	} `xml:"cve_items"`
}

// CVEParser handles parsing from multiple sources.
type CVEParser struct {
	config    Config
	cache     map[string]*CVE
	cacheMu   sync.RWMutex
	tempDir   string
	httpClient *http.Client
}

// NewCVEParser creates a new parser with optional config override.
func NewCVEParser(cfg ...Config) *CVEParser {
	if len(cfg) == 0 || cfg[0].BatchSize <= 0 {
		cfg = append(cfg, DefaultConfig())
	}
	return &CVEParser{
		config:    cfg[0],
		cache:     make(map[string]*CVE),
		tempDir:   cfg[0].TempDir,
		httpClient: &http.Client{Timeout: cfg[0].Timeout},
	}
}

// ParseNVDJSON parses NVD JSON format. Returns parsed CVEs and errors.
func (p *CVEParser) ParseNVDJSON(data []byte) ([]*CVE, error) {
	var nvd NVDJSON
	if err := json.Unmarshal(data, &nvd); err != nil {
		return nil, fmt.Errorf("unmarshal NVD JSON: %w", err)
	}

	now, _ := time.Parse(time.RFC3339, nvd.Metadata.UpdatedDate)

	var results []*CVE
	for _, item := range nvd.CVEItems {
		if !isValidState(item.State) {
			continue
		}

		cve, err := p.parseNVDItem(&item, now)
		if err != nil {
			p.config.LogLevel = "debug"
			return results, fmt.Errorf("parse item %s: %w", item.ID, err)
		}

		results = append(results, cve)
	}

	return results, nil
}

// parseNVDItem parses a single NVD CVE item.
func (p *CVEParser) parseNVDItem(item *struct {
	ID              string     `json:"cve_id"`
	State           string     `json:"state"`
	PublishedDate   string     `json:"published_date"`
	UpdatedDate     string     `json:"updated_date"`
	VendorIDs       []string   `json:"vendor_ids"`
	Desc            struct {
		Value    string `json:"value"`
		Language string `json:"lang,omitempty"`
	} `json:"desc"`
	Summary         string     `json:"summary"`
	CvssData        []struct {
		Version   string  `json:"version"`
		VectorStr string  `json:"vectorString"`
		Score      float64 `json:"score,omitempty"`
	} `json:"cvss_data"`
	AffectedProducts []struct {
		Vendor    string `json:"vendor"`
		Product   string `json:"product"`
		Version   string `json:"version"`
		CvssData  []struct {
			Version   string  `json:"version"`
			VectorStr string  `json:"vectorString"`
			Score      float64 `json:"score,omitempty"`
		} `json:"cvss_data"`
	} `json:"affected_products"`
	References []struct {
		URL       string `json:"url"`
		Type      string `json:"ref_source"`
		Description struct {
			Value    string `json:"value"`
			Language string `json:"lang,omitempty"`
		} `json:"desc"`
	} `json:"references"`
	Publishers []struct {
		Name       string `json:"name"`
		URL        string `json:"url"`
		Category   string `json:"category"`
		ContactURL string `json:"contact_url,omitempty"`
	} `json:"publishers"`
}, now) (*CVE, error) {

	cve := &CVE{
		ID:              item.ID,
		PublishedDate:   p.parseDateTime(item.PublishedDate),
		UpdatedDate:     p.parseDateTime(item.UpdatedDate),
		Summary:         item.Summary,
	}

	if item.Desc.Value != "" {
		cve.Description = item.Desc.Value
	}

	// Extract CVSS v3 data
	for _, cvss := range item.CvssData {
		if cvss.Version == "3.0" || cvss.Version == "3.1" {
			cve.CVSSv3Score = cvss.Score
			cve.CVSSv3Vector = cvss.VectorStr
			break
		}
	}

	// Extract CVSS v2 data
	for _, cvss := range item.CvssData {
		if cvss.Version == "2.0" || cvss.Version == "2.1" {
			cve.CVSSv2Score = cvss.Score
			cve.CVSSv2Vector = cvss.VectorStr
			break
		}
	}

	// Parse affected products and systems
	for _, prod := range item.AffectedProducts {
		if prod.Product == "" || prod.Version == "" {
			continue
		}

		entry := AffectedEntry{
			Product:       prod.Product,
			VersionRange:  prod.Version,
			PublishedDate: now,
		}

		for _, cvss := range prod.CvssData {
			if cvss.Version == "3.0" || cvss.Version == "3.1" {
				entry.CVSSv3Score = cvss.Score
				break
			} else if cvss.Version == "2.0" || cvss.Version == "2.1" {
				entry.CVSSv2Score = cvss.Score
				break
			}
		}

		cve.AffectedSystems = append(cve.AffectedSystems, entry)
	}

	// Parse references
	for _, ref := range item.References {
		if ref.URL == "" {
			continue
		}

		refEntry := Reference{
			URL:  ref.URL,
			Type: strings.ToLower(ref.Type),
		}

		if ref.Description.Value != "" {
			refEntry.Description = ref.Description.Value
		}

		cve.References = append(cve.References, refEntry)
	}

	// Parse publishers
	for _, pub := range item.Publishers {
		if pub.Name == "" || pub.URL == "" {
			continue
		}

		pubEntry := Publisher{
			Name:       pub.Name,
			URL:        pub.URL,
			Category:   strings.ToLower(pub.Category),
			ContactURL: pub.ContactURL,
		}

		cve.Publishers = append(cve.Publishers, pubEntry)
	}

	return cve, nil
}

// ParseCISAXML parses CISA XML format. Returns parsed CVEs and errors.
func (p *CVEParser) ParseCISAXML(data []byte) ([]*CVE, error) {
	var cisax CISAXML
	if err := json.Unmarshal(data, &cisax); err != nil {
		return nil, fmt.Errorf("unmarshal CISA XML: %w", err)
	}

	now, _ := time.Parse(time.RFC3339, cisax.Metadata.UpdatedDate)

	var results []*CVE
	for _, item := range cisax.CVEItems {
		if !isValidState(item.State) {
			continue
		}

		cve, err := p.parseCISAItem(&item, now)
		if err != nil {
			p.config.LogLevel = "debug"
			return results, fmt.Errorf("parse item %s: %w", item.ID, err)
		}

		results = append(results, cve)
	}

	return results, nil
}

// parseCISAItem parses a single CISA CVE item.
func (p *CVEParser) parseCISAItem(item *struct {
	ID              string     `xml:"cve_id"`
	State           string     `xml:"state"`
	PublishedDate   string     `xml:"published_date"`
	UpdatedDate     string     `xml:"updated_date"`
	VendorIDs       []string   `xml:"vendor_ids"`
	Desc